lib/VideoInputTest.ts

```diff
@@ -0,0 +1,372 @@
+ */
+export class VideoInputTest extends EventEmitter {
+  /**
+   * The name of the test.
+   */
+  static testName: string = 'video-input-test';
+  /**
+   * Default options for the test.
+   */
+  private static defaultOptions: VideoInputTest.InternalOptions = {
+    debug: false,
+    duration: Infinity,
+    getUserMedia,
+  };
+
+  /**
+   * Teststamp of when the test was ended.
+   */
+  private _endTime: number | null = null;
+  /**
+   * An array of any errors that occur during the run time of the test.
+   */
+  private readonly _errors: DiagnosticError[] = [];
+  /**
+   * Options to be used throughout the runtime of the test.
+   */
+  private _options: VideoInputTest.InternalOptions;
+  /**
+   * The promise returned by the `HTMLMediaElement` after playing the stream
+   * captured by `getUserMedia`.
+   */
+  private _playPromise: Promise<void> | null = null;
+  /**
+   * Teststamp of when the test was constructed and started.
+   */
+  private _startTime: number;
+  /**
+   * Timer ID of the test duration timeout.
+   */
+  private _timeoutId: NodeJS.Timeout | null = null;
+  /**
+   * The `MediaStream` resulting from calling `getUserMedia`.
+   */
+  private _userMediaStream: MediaStream | null = null;
+
+  /**
+   * Constructor for a [[VideoInputTest]] object.
+   * @param options Options to be used during the runtime of the test.
+   */
+  constructor(options?: VideoInputTest.Options) {
+    super();
+    this._startTime = Date.now();
+    this._options = { ...VideoInputTest.defaultOptions, ...options };
+
+    /**
+     * Use `setTimeout` to allow event listeners to properly bind before
+     * starting the test.
+     */
+    setTimeout(() => this._startTest());
+  }
+
+  /**
+   * Stops the test. Emits a report upon the end of the test.
+   *
+   * @event VideoInputTest.Events.End
+   */
+  stop(): void {
+    if (typeof this._endTime === 'number') {
+      this._onWarning(new AlreadyStoppedError());
+      return;
+    }
+
+    this._endTime = Date.now();
+
+    const {
+      width: streamWidth,
+      height: streamHeight,
+      deviceId,
+    } = this._userMediaStream?.getVideoTracks()[0].getSettings() || {};
+
+    const report: VideoInputTest.Report = {
+      deviceId,
+      errors: this._errors,
+      resolution: { width: streamWidth || 0, height: streamHeight || 0 },
+      testName: VideoInputTest.testName,
+      testTiming: {
+        duration: this._endTime - this._startTime,
+        end: this._endTime,
+        start: this._startTime,
+      },
+    };
+
+    this.emit(VideoInputTest.Events.End, report);
+
+    this._cleanup();
+  }
+
+  /**
+   * Clean up the test.
+   */
+  private _cleanup(): void {
+    if (this._userMediaStream) {
+      this._userMediaStream.getTracks().forEach((track: MediaStreamTrack) => {
+        track.stop();
+      });
+      if (this._options.element) {
+        this._options.element.srcObject = null;
+        if (this._playPromise) {
+          this._playPromise.then(() => {
```

Comment:
Not entirely sure, but I was under the impression that we would only need to `pause` if `play` successfully started, so only on `.then` instead of `.finally`. I'll look into this.